### PR TITLE
[RFC] Converting Active Record to the Data Mapper pattern

### PR DESCRIPTION
**Pull request by SidRoberts**
So I thought I'd have a go at converting Phalcon's Active Record models to the Data Mapper pattern (see [Doctrine](http://www.doctrine-project.org/projects/orm.html)) and I want to hear people's thoughts and opinions before I take it any further.

I've avoided doing to much work just in case it ends up being a dead-end so no code has been refactored and methods in the models manager now call protected methods in Phalcon\Mvc\Model, etc; - it's a bit of a mess. Obviously this will/should all change.

Put succinctly, methods have been moved to the models manager and (currently) operate exactly as they did before:

``` php
// Before
$person = People::findFirst($params);

// After
$person = $modelsManager->findFirst(
    People::class,
    $params
);
```

``` php
// Before
$person = People::count($params);

// After
$person = $modelsManager->count(
    People::class,
    $params
);
```

``` php
// Before
$person->delete();

// After
$modelsManager->delete($person);
```

``` php
// Before
$person->save($params);

// After
$modelsManager->save($person, $params);
```

Phalcon\Mvc\Model is still far too complicated, in my opinion, so there's still a lot of work to do. :stuck_out_tongue: 

A few things that I'd like to do is remove the ability to assign properties during create/update/save:

``` php
$robot->save(
    [
        "name" => "Bender",
    ]
);

$modelsManager->save(
    $robot,
    [
        "name" => "Bender",
    ]
);
```

I know people like it because it acts a little shortcut but it doesn't make sense in the data mapper context and looks a little clumsy. A method called `save()` should save and do nothing else.

I'd also like to decide between only supporting [public properties or getters/setters](https://docs.phalconphp.com/en/latest/reference/models.html#public-properties-vs-setters-getters) but have no personal preference of either. Alternatively (or additionally), make an effort to vastly simplify how properties are get/set internally.

(It goes without saying but is obviously aimed at version 4 - not version 3)

Thoughts?

### CODE REVIEW

**Comment by Jurigag** [comment]
Intorduce pls repository class imho as well, because using modelsManager is dirty. So we can still jus t do `$personRepostory->find();` for example.

Also this change is of course for 4.0.0. I thought @andresgutierrez was planning this for 4.0.0

**Review by sergeyklay — APPROVED** [review]
Looks great to me

File: phalcon/mvc/model/manager.zep [hunk 1]
```diff
@@ -1805,4 +1811,888 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
 		Query::clean();
 	}
+
+
+
+
+
+
+
+
```
Comment (Green-Cat) [inline phalcon/mvc/model/manager.zep]: Too much spacing.

File: phalcon/mvc/model/manager.zep [hunk 2]
```diff
@@ -1805,4 +1811,888 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+				resultset->setHydrateMode(hydration);
+			}
+		}
+
+		return resultset;
+	}
+
+
+	/**
+	 * Allows to query the first record that match the specified conditions
```
Comment (Green-Cat) [inline phalcon/mvc/model/manager.zep]: Allows to query for the first record that matches the specified conditions

File: phalcon/mvc/model/manager.zep (hunk 1 above)
Comment (SidRoberts) [inline phalcon/mvc/model/manager.zep]: I was going for speed. :wink:

Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: No need for speed, it will be at least few months until we create 4.0.x repo i guess, and few years to release :D

**Comment by SidRoberts** [comment]
@Jurigag, it'd be tedious to have to create repository classes for each model. We could set the model class name in the constructor though:

```zephir
namespace Phalcon\Mvc\Model;

class Repository
{
	public function __construct(string! modelClass);

	public function find(array params = []) -> <ResultsetInterface>;

	public function findFirst(array params = []) -> <ModelInterface>;

	public function count(array parameters = []) -> int;

	public function sum(array parameters = []);

	public function maximum(array parameters = []);

	public function minimum(array parameters = []);

	public function average(array parameters = []);
}
```

It would reduce the complexity of the Models Manager a bit but would repositories add complexity for  users/developers?:

```php
$robots = $modelsManager->findFirst(
    Robots::class,
    $params
);

// OR ...

$robotsRepository = $modelsManager->getRepository(
    Robots::class
);

$robots = $robotsRepository->findFirst($params);
```

File: phalcon/mvc/model/manager.zep (hunk 2 above)
Comment (SidRoberts) [inline phalcon/mvc/model/manager.zep]: Fixed in 77acbfe.

**Comment by Jurigag** [comment]
I mean if you don't want - you don't need to create repository. You can use as well just `$modelsManager->findFirst(Robots::class, $params)` i guess maybe. Repository will be just some kind of shorthand and some place where you can add addtional methods for finding stuff. I know developer can always create his own method but in most Data Mappers there is Repository class introduced.

**Review by Jurigag — CHANGES_REQUESTED** [review]
`Phalcon/Mvc/Model/Repository` class, add option to use custom Repository class for models.

File: phalcon/mvc/model/manager.zep [hunk 3]
```diff
@@ -1805,4 +1811,855 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+		 * Fire the start event
+		 */
+		model->fireEvent("prepareSave");
+
+		/**
+		 * Save related records in belongsTo relationships
+		 */
+		let related = model->_related;
+		if typeof related == "array" {
+			if model->_preSaveRelatedRecords(writeConnection, related) === false {
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: There should be something like `Phalcon\Mvc\Model\Worker` or `Phalcon\Mvc\Model\UnitOfWork` which will do this.

File: phalcon/mvc/model/manager.zep [hunk 4]
```diff
@@ -1805,4 +1811,855 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+			}
+
+			return false;
+		}
+
+		/**
+		 * Depending if the record exists we do an update or an insert operation
+		 */
+		if exists {
+			let success = model->_doLowUpdate(metaData, writeConnection, table);
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: There should be something like `Phalcon\Mvc\Model\Worker` or `Phalcon\Mvc\Model\UnitOfWork` which will do this.

File: phalcon/mvc/model/manager.zep [hunk 5]
```diff
@@ -1805,4 +1811,855 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+			/**
+			 * Rollbacks the implicit transaction if the master save has failed
+			 */
+			if success === false {
+				writeConnection->rollback(false);
+			} else {
+				/**
+				 * Save the post-related records
+				 */
+				let success = model->_postSaveRelatedRecords(writeConnection, related);
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: There should be something like `Phalcon\Mvc\Model\Worker` or `Phalcon\Mvc\Model\UnitOfWork` which will do this.

File: phalcon/mvc/model/manager.zep [hunk 6]
```diff
@@ -1805,4 +1811,855 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+		 * _postSave() invokes after* events if the operation was successful
+		 */
+		if globals_get("orm.events") {
+			let success = model->_postSave(success, exists);
+		}
+
+		if success === false {
+			model->_cancelOperation();
+		} else {
+			model->fireEvent("afterSave");
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: There should be something like `Phalcon\Mvc\Model\Worker` or `Phalcon\Mvc\Model\UnitOfWork` which will do this.

File: phalcon/mvc/model/manager.zep [hunk 7]
```diff
@@ -1805,4 +1811,855 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+
+		/**
+		 * _postSave() invokes after* events if the operation was successful
+		 */
+		if globals_get("orm.events") {
+			let success = model->_postSave(success, exists);
+		}
+
+		if success === false {
+			model->_cancelOperation();
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: There should be something like `Phalcon\Mvc\Model\Worker` or `Phalcon\Mvc\Model\UnitOfWork` which will do this.

**Review by Jurigag — COMMENTED** [review]
Worker class for model methods:
- _preSaveRelatedRecords
- _exists
- _preSave
- _doLowUpdate
- _doLowInsert
- _postSaveRelatedRecords
- _postSave
- _cancelOperation
- fireEvent
- _checkForeignKeysReverseRestrict
- _checkForeignKeysReverseCascade

So we can remove all this not needed logic in model. It should impelemnt `EvensAwareInterface` so we can attach some listener to it. Not sure about which events exactly - i guess preSave, postSave, onCancelOperation, onUpdate, onInsert ? Also as well existing model events so we can watch all model events fired for all models in one listener.

**Review by Jurigag — CHANGES_REQUESTED** [review]
Test cases for class implementing only ModelInterface not extending Model,  in future for any class without any interface or extending class.

**Comment by SidRoberts** [comment]
What uses could a custom repository have? In my head, I imagine Repository only has [the methods listed above](https://github.com/phalcon/cphalcon/pull/12317#issuecomment-253839204) (as well as support for findFirstBy..., findBy..., countBy..., etc;).

It might be a good idea to allow the developer to overwrite the repository class via the DI:

```php
$di->set("modelRepository", "Phalcon\\Mvc\\Model\\Repository");

// OR...

$di->set("Phalcon\\Mvc\\Model\\Repository", "MyCustomRepositoryClass");
```

```zephir
let repository = <RepositoryInterface> dependencyInjector->get(
    "modelRepository",
    [modelClass, this]
);

// OR...

let repository = <RepositoryInterface> dependencyInjector->get(
    "Phalcon\\Mvc\\Model\\Repository",
    [modelClass, this]
);
```

**Comment by Jurigag** [comment]
Repository also should provide query builder property(`Phalcon\Mvc\Model\Query\Builder`) and query method(`Phalcon\Mvc\Model`Criteria`). With builder we can do extensive queries. So there is example:

```php
class PersonRepository extends Repository
{
    public function findWithProfiles()
    {
        return $this->queryBuilder->from('Person')
            ->columns('*')
            ->leftJoin('Profile')
            ->getQuery()
            ->execute();
    }
}
```

**Comment by Jurigag** [comment]
@SidRoberts 

Also this is really great idea. But this definitely need a hard work and implement new classes. It's not only change a few we have right now = and we have data mapper. I just add my ideas what i think should be here and how it should look. How it will be implemented and what's your idea it's up to you and someone who will merge this.

**Comment by SidRoberts** [comment]
It's been a long day and after spending so much time looking at code, I can't see straight anymore. I need a good night's sleep before I can do any more serious coding.  :stuck_out_tongue_closed_eyes: I agree with your ideas about custom repositories though.

How about copying how the Models Manager deals with custom model sources and schemas (`Model::setSource()` / `Model::setSchema()`) and managing custom repositories like this:

```php
class Robots extends \Phalcon\Mvc\Model
{
	public function initialize()
	{
		$this->setRepository(
			RobotsRepository::class
		);
	}
}
```

I think I'll use this PR as a playground for ideas and redo the entire thing when the 4.0.x branch is created and everyone's opinions are taken into account.

**Comment by Jurigag** [comment]
Yea This is good way BUT i'm not sure about $this->setRepository. DataMapper is all about NOT EXTENDING any class i think.

I think in result and in final form the only needed thing for our models in phalcon 4.0.0 should be a) implement interface b) use trait c) don't use anything. Im just not sure about solution. For NOW your solution is fine, but i think we should avoid ` extends \Phalcon\Mvc\Model` and in final form it should aim to not having ` extends \Phalcon\Mvc\Model` and somethign like `use Model` as a trait(as far as i know traits are just classes internally ? Correct me if im wrong. Or just only some interface for needed things to work. And repository would be just `public static function getRepository()`

**Comment by SidRoberts** [comment]
I'm glad you're here - you're right and I'm thinking of solutions that require the least amount of effort to implement. :wink: 

I've never used traits but I seem to remember that it's akin to copying and pasting chunks of code, as far as the compiler is concerned.

We'll figure it out tomorrow.

**Comment by Jurigag** [comment]
But im not sure, maybe there should be like totally nothing, this is up to people with more knowledge, don't if DataMapper pattern allows anything like trait or something like this. Well in doctrine there are annotations, and something is parsing them, so well - this is just some kind of "workaround" in doctrine to not having trait or something like this.

**Comment by SidRoberts** [comment]
That's kind of the problem. Doctrine stores a lot of metadata in annotations and I don't think annotations will be considered fast enough. We could have metadata stored somewhere else (YAML, XML, ...) but that'd be messy. Or we could just have a few getters in Model.

**Comment by Jurigag** [comment]
Or just trait and still use $this->xyz but without extends.

**Comment by Jurigag** [comment]
Well it's not so easy, but maybe this is a FINALLY time to introduce traits to zephir ? @andresgutierrez @sergeyklay we are waiting already 2 years for it https://github.com/phalcon/zephir/issues/504

**Comment by Jurigag** [comment]
Well i will add trait support for PHP(only atm) in zephir soon so DataMapper will work as trait in php classes.

File: phalcon/mvc/model/repository.zep [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+ */
+class Repository implements RepositoryInterface
+{
+
+	/**
+	 * @var string
+	 */
+	protected _modelClass;
+
+	protected _modelsManager;
```
Comment (Jurigag) [inline phalcon/mvc/model/repository.zep]: Oh okay, there is already `_modelsManager`, forgot what i wrote about `createBuilder` in repository then :D

File: phalcon/mvc/model/manager.zep [hunk 9]
```diff
@@ -1805,4 +1810,459 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
...
+			model->fireEvent("afterSave");
+		}
+
+		return success;
+	}
+
+	/**
+	 * Executes internal events after save a record
+	 */
+	protected function _postSave(<ModelInterface model, boolean success, boolean exists) -> boolean
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: Don't agree with having this in manager imho. As i wrote already we should introduce some new class.

**Comment by Jurigag** [comment]
I don't like making this manager such a big class now, we definitely need new class what i already wrote imho.

File: phalcon/mvc/model/manager.zep (hunk 9 above)
Comment (SidRoberts) [inline phalcon/mvc/model/manager.zep]: But UnitOfWork is effectively a transaction - it only makes sense to use it if multiple models are being saved at once. I want to get all the pre/post protected methods into Manager just for now so they're all together and we can move again them later.

**Comment by SidRoberts** [comment]
I've reverted the last few commits. There was a bug somewhere in there anyway. :stuck_out_tongue:

File: phalcon/mvc/model/manager.zep [hunk 10]
```diff
@@ -2249,4 +2235,157 @@ class Manager implements ManagerInterface, InjectionAwareInterface, EventsAwareI
 
 		return success;
 	}
+
+	/**
+	 * Checks if a record already exists or not
+	 */
+	public function exists(<ModelInterface> model) -> boolean
```
Comment (Jurigag) [inline phalcon/mvc/model/manager.zep]: Well this can be in manager maybe, but i think the rest of `_anyMethod` from model should go to new class.

**Comment by allflame** [comment]
@Jurigag on a side note: why not use doctrine as is?

**Comment by Jurigag** [comment]
Because doctrine is PHP and phalcon is c and is all about performance and doctrine is slowest ORM available for php ?

**Comment by allflame** [comment]
If you look into the comparison you will see a strong correlation between UoW/Datamapper (slower) - ActiveRecord (faster).
I do understand the benefits of writing Zephir code for the Phalcon ORM and maybe this is a good idea.

**Comment by Jurigag** [comment]
Well this is why we have zephir and we cant create something which will be still fast enough and will get all advantages of Datamapper - but it should be a real Datamapper or be really similar to it.

**Comment by sergeyklay** [comment]
abandoned project https://github.com/lynx/lynx-legacy-zephir

**Comment by Jurigag** [comment]
@SidRoberts also if you will return to this it could be nice to add some event like notFound or something if result of `findFirst` don't exist in database

**Comment by SidRoberts** [comment]
I'm currently working on moving the relevant unit tests to Codeception and tidying them up a bit (#12348), refactoring model classes (#12328) and doing anything else to reduce the possibility of messy merge conflicts and introducing regressions.

**Comment by SidRoberts** [comment]
How about introducing this slowly over the course of version 3, keeping BC along the way and then drop BC or deprecate in version 4? Or would that be too complicated?

**Comment by Jurigag** [comment]
Too complicated in my opinion. Better just create 4.0.x branch and merge it while it's be done and that's it.

**Comment by tempcke** [comment]
I've been doing a lot of reading lately about data mapper vs active record and the more I read the more I like the data mapper pattern.  Though I do not like doctrine due to all the dang @annotations.  Is it your goal to replace active record with data mapper pattern or will both be available ?  Will the domain models be required to extend a base class?

**Comment by virgofx** [comment]
Just a note theres no way the active record can be replaced in Phalcon as this is one of the core features of the framework. Any new ORM will need to be an additional option that doesn't impact existing active record.

**Comment by SidRoberts** [comment]
I completely messed up the history and there were too many merge conflicts to deal with so I'm recreating these commits manually. The original branch is available at https://github.com/SidRoberts/cphalcon/commits/v4-model-data-mapper-original

**Comment by sergeyklay** [comment]
@SidRoberts Could you please check tests

**Comment by sergeyklay** [comment]
Thank you

**Comment by MufidJamaluddin** [comment]
I think the better option is move pagination method to this repository or make new class (PagingRepository)

**Pull request merged**